core/logger/splunk.go

```diff
@@ -0,0 +1,153 @@
+	Index      string          `json:"index,omitempty"`      // optional name of the Splunk index to store the event in; not required if the token has a default index set in Splunk
+	Event      json.RawMessage `json:"event"`                // zap json logger
+}
+
+type splunkSink struct {
+	HTTPClient    httpClient
+	nowFn         func() int64
+	URL           string // URL to Splunk HTTP Event Collector
+	Hostname      string // Hostname is the human-readable machine name sending logs
+	Token         string // Splunk HTTP Event Collector token
+	Source        string // Splunk source field value, description of the source of the event
+	SourceType    string // Splunk source type, optional name of a sourcetype field value
+	Index         string // Splunk Index, optional name of the Splunk index to store the event in
+	SkipTLSVerify bool   // Skip verifying the certificate of the HTTP Event Collector
+	events        [][]byte
+	sync.Mutex
+}
+
+func envVarOrDefault(key string, defaultValue string) string {
+	envVar := os.Getenv(key)
+	if envVar == "" {
+		return defaultValue
+	}
+	return envVar
+}
+
+func initialize(s *splunkSink) *splunkSink {
+	tr := &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: s.SkipTLSVerify}}
+	s.HTTPClient = &http.Client{Timeout: time.Second * 20, Transport: tr}
+
+	if s.Hostname == "" {
+		s.Hostname, _ = os.Hostname()
+	}
+	if s.nowFn == nil {
+		s.nowFn = func() int64 {
+			return time.Now().UnixNano()
+		}
+	}
+
+	return s
+}
+
+func newSplunkSink() (*splunkSink, error) {
+	s := &splunkSink{}
+	s.URL = os.Getenv("SPLUNK_URL")
+	s.Source = envVarOrDefault("SPLUNK_SOURCE", "chainlink")
+	s.SourceType = envVarOrDefault("SPLUNK_SOURCETYPE", "chainlink")
+	s.Index = os.Getenv("SPLUNK_INDEX")
+	s.Hostname = os.Getenv("SPLUNK_HOSTNAME")
+	s.Token = os.Getenv("SPLUNK_TOKEN")
+	s.SkipTLSVerify, _ = strconv.ParseBool(os.Getenv("SPLUNK_SKIP_TLS_VERIFY"))
+
+	return initialize(s), nil
+}
+
+func (p *splunkSink) Close() error {
+	return nil
+}
+
+// Write implement zap.Sink func Write
+func (p *splunkSink) Write(b []byte) (n int, err error) {
+	e := &event{
+		Time:       p.nowFn(),
+		Host:       p.Hostname,
+		Source:     p.Source,
+		SourceType: p.SourceType,
+		Index:      p.Index,
+		Event:      json.RawMessage(b),
+	}
+	encoded, err := json.Marshal(e)
+	if err != nil {
+		return 0, err
+	}
+	p.Lock()
+	defer p.Unlock()
+	p.events = append(p.events, encoded)
+	return len(b), nil
+}
+
+// Sync implement zap.Sink func Sync
+func (p *splunkSink) Sync() error {
+	eventsLen := len(p.events)
+	if eventsLen > 0 {
+		p.Lock()
+		eventsToLog := p.events[0:eventsLen]
+		p.events = p.events[eventsLen:]
+		p.Unlock()
+		return p.logEvents(eventsToLog)
+
+	}
+	return nil
+}
+
+func (p *splunkSink) logEvents(events [][]byte) error {
+	buf := new(bytes.Buffer)
+	for _, e := range events {
+		buf.Write(e)
+	}
+
+	return p.doRequest(buf)
+}
+
+func (p *splunkSink) doRequest(b *bytes.Buffer) error {
```

Comment:
Does this synchronously write every log event over HTTP?

Reply:
This is synchronous, but is triggered by Sync(), which is most of the time asynchronous, except when called direct by your main code to sync logs before closing the application.